src/components/appointments/AppointmentFields.js

```diff
@@ -0,0 +1,189 @@
+import React, {useContext, useState} from 'react';
+import {FormattedMessage, useIntl} from 'react-intl';
+import {useAsync} from 'react-use';
+
+import {ConfigContext} from 'Context';
+import {get} from 'api';
+import Label from 'components/Label';
+import Loader from 'components/Loader';
+import {getBEMClassName, getFormattedDateString, getFormattedTimeString} from 'utils';
+
+const requestOrEmpty = async (url, query, condition) => {
+  if (condition) {
+    return await get(url, query);
+  } else {
+    return [];
+  }
+};
+
+const ProductField = ({value, onChange, url}) => {
+  const {loading, value: items} = useAsync(async () => await get(url));
+  // todo process errors
+
+  return (
+    <div className={getBEMClassName('form-control')}>
+      <Label>
+        <FormattedMessage description="Appointment product field label" defaultMessage="Product" />
+      </Label>
+
+      {loading ? (
+        <Loader />
+      ) : (
+        <select className={getBEMClassName('select')} value={value} onChange={onChange}>
+          <option value={''}></option>
+          {items.map(item => (
+            <option value={item.identifier} key={item.identifier}>
+              {item.name}
+            </option>
+          ))}
+        </select>
+      )}
+    </div>
+  );
+};
+
+const LocationField = ({value, onChange, url, urlQuery}) => {
+  const {loading, value: items} = useAsync(
+    async () => await requestOrEmpty(url, urlQuery, urlQuery.product_id),
+    [urlQuery.product_id]
+  );
+
+  return (
+    <div className={getBEMClassName('form-control')}>
+      <Label>
+        <FormattedMessage
+          description="Appointment location field label"
+          defaultMessage="Location"
+        />
+      </Label>
+
+      {loading ? (
+        <Loader />
+      ) : (
+        <select className={getBEMClassName('select')} value={value} onChange={onChange}>
+          <option value={''}></option>
+          {items.map(item => (
+            <option value={item.identifier} key={item.identifier}>
+              {item.name}
+            </option>
+          ))}
+        </select>
+      )}
+    </div>
+  );
+};
+
+const DateField = ({value, onChange, url, urlQuery}) => {
+  const intl = useIntl();
+  const {loading, value: items} = useAsync(
+    async () => await requestOrEmpty(url, urlQuery, urlQuery.product_id && urlQuery.location_id),
+    [urlQuery.product_id, urlQuery.location_id]
+  );
+
+  return (
+    <div className={getBEMClassName('form-control')}>
+      <Label>
+        <FormattedMessage description="Appointment date field label" defaultMessage="Date" />
+      </Label>
+
+      {loading ? (
+        <Loader />
+      ) : (
+        <select className={getBEMClassName('select')} value={value} onChange={onChange}>
+          <option value={''}></option>
+          {items.map(item => (
+            <option value={item.date} key={item.date}>
+              {getFormattedDateString(intl, item.date)}
```

Comment:
we prefer using the `react-intl` `FormattedDate` component - the `useIntl` imperative API is an escape hatch for when we're outside of React context.

Reply:
I tried to use `<FormattedDate/>` and `<FormattedTime/>` in option labels, but the values were not shown in the dropdowns. So I left it as it is